src/java.base/share/classes/java/math/MutableBigInteger.java

```diff
@@ -715,14 +738,28 @@ private final void primitiveRightShift(int n) {
      * Assumes that intLen > 0, n > 0 for speed
      */
     private final void primitiveLeftShift(int n) {
+        primitiveLeftShift(n, value, offset);
+    }
+
+    /**
+     * Left shift this MutableBigInteger n bits, where n is
+     * less than 32, placing the result in the specified array.
+     * Assumes that intLen > 0, n > 0 for speed.
+     * The result can be the value array of this MutableBigInteger,
+     * but for speed the copy is not performed safely, so, in that case
+     * the caller has to make sure that
+     * {@code (resPos <= offset || resPos >= offset + intLen)}.
+     */
+    private final void primitiveLeftShift(int n, int[] result, int resPos) {
```

Comment:
I think this method can be made more symmetrical w.r.t. `primitiveRightShift()` if starting from the right (least significant `int`).

Reply:
Yes, it could, but the problem is that in this way the precondition `(resPos <= offset || resPos >= offset + intLen)` is no longer correct, and in particular `resPos <= offset` is used by `MBI.leftShift()` if the `result == value`.

Reply:
I mean something like
```
    private void primitiveRightShift(int n, int[] result, int resPos) {
        int[] val = value;
        int n2 = 32 - n;
        int c = 0;
        for (int i = 0; i < intLen; i++) {
            int b = val[offset + i];
            result[resPos + i] = c << n2 | b >>> n;
            c = b;
        }
    }
```
and
```
    private void primitiveLeftShift(int n, int[] result, int resPos) {
        int[] val = value;
        int n2 = 32 - n;
        int c = 0;
        for (int i = intLen - 1; i >= 0; i--) {
            int b = val[offset + i];
            result[resPos + i] = c >>> n2 | b << n;
            c = b;
        }
```
They are compatible with the precondition, right?

Reply:
But in the second version, the precondition must be `resPos <= offset - intLen  || resPos >= offset` to make sure that the result is correct whether `result == value`, and the condition `resPos <= offset - intLen` is stronger than `resPos <= offset`.

Reply:
Yes, I agree.

This one should correctly account for the precondition, and IMO is slightly simpler to read.
```
    private void primitiveLeftShift(int n, int[] result, int resPos) {
        int[] val = value;
        int n2 = 32 - n;
        final int m = intLen - 1;
        int b = val[offset];
        for (int i = 0; i < m; i++) {
            int c = val[offset + i + 1];
            result[resPos + i] = (b << n) | (c >>> n2);
            b = c;
        }
        result[resPos + m] = b << n;
    }
```